traits/trait_types.py

```diff
@@ -2774,6 +2774,169 @@ def items_event(self):
 AdaptMap = {"no": 0, "yes": 1, "default": 2}
 
 
+class Map(TraitType):
+    """ Checks that the value assigned to a trait attribute is a key of a
+        specified dictionary, and also assigns the dictionary value
+        corresponding to that key to a *shadow* attribute.
+
+        A trait attribute of type Map is called a *mapped* trait
+        attribute. In practice, this means that the resulting object actually
+        contains two attributes: one whose value is a key of the Map
+        dictionary, and the other whose value is the corresponding value of the
+        Map dictionary. The name of the shadow attribute is simply the base
+        attribute name with an underscore ('_') appended. Mapped trait
+        attributes can be used to allow a variety of user-friendly input values
+        to be mapped to a set of internal, program-friendly values.
+
+        Example
+        -------
+
+        The following example defines a ``Person`` class::
+
+            >>>class Person(HasTraits):
+            ...    married = Map({'yes': 1, 'no': 0 }, default_value="yes")
+            >>>
+            >>>bob = Person()
+            >>>print(bob.married)
+            yes
+            >>>print(bob.married_)
+            1
+
+        In this example, the default value of the ``married`` attribute of the
+        Person class is 'yes'. Because this attribute is defined using
+        Map, instances of Person have another attribute,
+        ``married_``, whose default value is 1, the dictionary value
+        corresponding to the key 'yes'.
+
+        Parameters
+        ----------
+        map : dict
+            A dictionary whose keys are valid values for the trait attribute,
+            and whose corresponding values are the values for the shadow
+            trait attribute.
+
+        Attributes
+        ----------
+        map : dict
+            A dictionary whose keys are valid values for the trait attribute,
+            and whose corresponding values are the values for the shadow
+            trait attribute.
+    """
+
+    is_mapped = True
+
+    def __init__(self, map, **metadata):
+
+        self.map = map
+        self.fast_validate = (ValidateTrait.map, map)
+
+        default_value = metadata.pop("default_value", None)
+
+        if default_value is not None:
+            default_value = self.validate(None, None, default_value)
```

Comment:
I had this comment before, but I removed it before submitting the review, thinking I was wrong - now I wanted to raise this question again:

Speaking of consistency, I don't know of a subclass of `TraitType` that actually validates default value on init. e.g. this does not raise:
```
from traits.api import Int
a = Int("ABC")
```
Not that I think validating the default value is wrong per se. It is just that the validation here sets itself apart from the other trait types.

There might also be a reason why default values are never validated when the trait is initialized?

Could we remove this validation here (and for `PrefixMap`) for consistency and open an issue for deciding whether we should validate default values more generally?

Reply:
Yes, I've been validating default values for any new trait types I created. But we haven't had a discussion regarding this.

Reply:
I agree that it makes sense to stick to the current convention of not validating defaults, at least for now.

Reply:
Will update this thread after discussion.

Reply:
Opened #946 for deciding whether we should do validation generally (if we don't, at least we can track the reasons there).